modules/packages/UnrolledLinkedList.chpl

```diff
@@ -0,0 +1,1249 @@
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This module contains the implementation of the unrolledLinkedList type.
+
+  An unrolled linked list is a linked list of small arrays, all of the same size 
+  where each is so small that the insertion or deletion is fast and quick, but 
+  large enough to fill the cache line. 
+*/
+module UnrolledLinkedList {
+  import ChapelLocks;
+  private use HaltWrappers;
+  private use Sort;
+  private use IO;
+  private use List;
+
+  pragma "no doc"
+  private param _sanityChecks = true;
+
+  //
+  // Some asserts are useful while developing, but can be turned off when the
+  // implementation is correct.
+  //
+  private inline proc _sanity(expr: bool) {
+    if _sanityChecks then
+      assert(expr);
+  }
+
+  pragma "no doc"
+  type _lockType = ChapelLocks.chpl_LocalSpinlock;
+
+  //
+  // Use a wrapper class to let unrolledLinkedList methods have a const ref receiver even
+  // when `parSafe` is `true` and the unrolledLinkedList lock is used.
+  //
+  pragma "no doc"
+  class _LockWrapper {
+    var lock$ = new _lockType();
+
+    inline proc lock() {
+      lock$.lock();
+    }
+
+    inline proc unlock() {
+      lock$.unlock();
+    }
+  }
+
+  /* Check that element type is supported by unrolledLinkedList */
+  pragma "no doc"
+  proc _checkType(type eltType) {
+    if !isDefaultInitializable(eltType) {
+      compilerError("unrolledLinkedList element type must be default-initializable");
+    }
+  }
+
+  pragma "no doc"
+  proc _checkNodeCapacity(cap: int) {
+    if cap < 2 {
+      halt("nodeCapacity can't be less than 2");
+    }
+    if cap % 2 == 1 {
+      halt("nodeCapacity can't be odd");
+    }
+  }
+
+  pragma "no doc"
+  class _linkedNode {
+    type eltType;
+    const capacity: int = 32;
+    var size: int = 0;
+    var data: [0..#capacity] eltType;
+    var next: unmanaged _linkedNode(eltType)? = nil;
+
+    proc append(in x: eltType) lifetime this < x {
+      _sanity(size != capacity);
+      data[size] = x;
+      size += 1;
+    }
+
+    proc append(ref x: eltType) where isOwnedClass(x)
+    lifetime this < x {
+      _sanity(size != capacity);
+      data[size] = x;
+      size += 1;
+    }
+  };
+
+  record unrolledLinkedList {
+
+    /* The type of the elements contained in this unrolledLinkedList. */
+    type eltType;
+
+    /* If `true`, this unrolledLinkedList will perform parallel safe operations. */
+    param parSafe = false;
+
+    /* The capacity of one linked node in the unrolledLinkedList */
+    var nodeCapacity: int = 32;
+
+    pragma "no doc"
+    var _size = 0;
```

Comment:
Basically number of items in the list, right? Can you add a comment saying that?